src/core/credGrainView.js

```diff
@@ -54,77 +60,133 @@ export type ParticipantsOptions = {|
 export class CredGrainView {
   _credGraph: CredGraph;
   _ledger: Ledger;
-  _defaultTimeScope: TimeScopedCredGrainView;
+  _participants: Array<ParticipantCredGrain>;
+  _intervals: IntervalSequence;
 
   constructor(credGraph: CredGraph, ledger: Ledger) {
     this._credGraph = credGraph;
     this._ledger = ledger;
-    this._defaultTimeScope = new TimeScopedCredGrainView(
-      credGraph,
-      ledger,
-      -Infinity,
-      Infinity
-    );
+    this._participants = [];
+    this._intervals = credGraph.intervals();
+
+    const graphParticipants = new Map<IdentityId, GraphParticipant>();
+    for (const participant of credGraph.participants()) {
+      graphParticipants.set(participant.id, participant);
+    }
+    for (const account of ledger.accounts()) {
+      const graphParticipant = graphParticipants.get(account.identity.id);
+      if (!graphParticipant)
+        throw new Error(
+          "The graph is missing an account that exists in the ledger."
+        );
+
+      this._participants.push({
+        identity: account.identity,
+        cred: graphParticipant.cred,
+        credPerInterval: graphParticipant.credPerInterval,
+        grainEarned: account.paid,
+        grainEarnedPerInterval: this._calculateGrainEarnedPerInterval(account),
+      });
+    }
+  }
+
+  _calculateGrainEarnedPerInterval(account: Account): $ReadOnlyArray<Grain> {
+    const grainEarnedPerInterval = [];
+    for (const interval of this._intervals) {
+      const grainEarnedThisInterval =
+        account.allocationHistory.reduce((grain, allocationReceipt) => {
+          if (!grain) grain = G.ZERO;
+          if (
+            interval.startTimeMs < allocationReceipt.credTimestampMs &&
+            allocationReceipt.credTimestampMs <= interval.endTimeMs
+          )
+            grain = G.add(grain, allocationReceipt.grainReceipt.amount);
+        }) || G.ZERO;
+      grainEarnedPerInterval.push(grainEarnedThisInterval);
+    }
+    return grainEarnedPerInterval;
   }
 
   withTimeScope(
-    startTimeMs: number,
-    endTimeMs: number
+    startTimeMs: TimestampMs,
+    endTimeMs: TimestampMs
   ): TimeScopedCredGrainView {
     return new TimeScopedCredGrainView(
-      this._credGraph,
-      this._ledger,
+      this._participants,
+      this._intervals,
       startTimeMs,
       endTimeMs
     );
   }
 
   intervals(): IntervalSequence {
-    return this._defaultTimeScope.intervals();
+    return cloneDeep(this._intervals);
```

Comment:
Interesting that we clone the IntervalSequence but not the participants. If we're worried about a bad caller mutating things, wouldn't we also be worried that they may mutate the participant array?

I would probably prefer deepFreeze to cloneDeep but don't feel strongly.

Reply:
yeah, probably overkill